WAF

cataphract opened the pull request:
Adds WAF support and changes several aspects of the build process. The "build images" are gone; there is only one build image for each arch. The build image is musl/libc++ based and works on both musl and glibc. Increases the compiler requirements to C++20, since libddwaf requires that.

dmehala reviewed (commented):
I started to review 2-3 days ago under the assumption that the code was feature-complete. However, as I delved deeper, I found missing features, like HTTP/JSON template. It appears that new features are still being introduced, giving the impression that this should be treated as a draft PR. 

I have not tested it but good job. However, I've notice a fundamental issue with the usage of `single_trace`, but that's probably because you are still actively working on the WAF implementation. 

Here is a comprehensive list of areas we should address before delivering:
- Dependency on `rapidjson`.
- Clear separation between Tracing and WAF: Comment related to `RequestTracing`.
- Thread Pool usage: Unless I'm mistaken, we should avoid created a thread pool per `datadog_waf_thread_pool_name` directives.
- Thorough configuration testing: More extensive testing is necessary.
- `OWNERSHIP` file.
- Build
  - Introduce build argument to build either `Tracing` or `Tracing + ASM`.
  - Generate a single `.so` for ease of use for the customer.
- Enhanced observability: Add a span on `on_access` including details on the config file used, the rule executed, client ip and all WAF tags?

I refrained from reviewing anything under `securty/` because, ownership and potential extensive discussions on C++. Nevertheless, I recommend using comprehensible variable and function names and to stay consistent with the existing code.

Lastly, it's worth noting that while you've successfully generalized Remote Config, its usage here introduces an avoidable dependency. We should consider delivering first WAF without Remote Config, then, adding this feature later.

cataphract commented on .gitignore:
```diff
@@ -1,6 +1,3 @@
-# source code fetched by the build
-/nginx/
-
```
nginx is now fetched by cmake into the build dir so it works on mac os. The problem now is that nginx is configured on mac os, but then when it runs in docker it's on linux, so the build fails

dmehala replied:
Ah, I made a recent change in build #66. Now, CMake expects the NGINX source directory as input, specified with `-DNGINX_SRC_DIR`, and configures accordingly. Instead of downloading the source code each time, which isn't ideal given that I work with both the stable and mainline NGINX branches, there are a few advantages to this approach:

1. It's quicker to simply rename the folder or adjust the project configuration to use a different directory rather than waiting for the script to download.
2. By avoiding unnecessary downloads every-time consider it should re-configure the project, I'm contributing to saving the planet.
3. I'm not aware of a .tar.gz file following the HEAD of NGINX.

At head it should work on MacOS. If not, that's something we can discuss.

cataphract commented on .gitmodules:
```diff
@@ -1,3 +1,6 @@
 [submodule "dd-trace-cpp"]
 	path = dd-trace-cpp
 	url = ../dd-trace-cpp.git
+[submodule "rapidjson"]
+	path = rapidjson
+	url = git@github.com:Tencent/rapidjson.git
```
The problem without that is that libddwaf is usually consumed in binary form (the exception being PHP), so it only provides a C interface.

While we could provide a C interface for this, and even return a pair<ddwaf_object* root, arena>, it's not that useful as a general matter. For instance, take Java. Using that native interface would involve converting the whole json string from a java string into a UTF-8 string (from UTF-16) and then do a native call, whereas we can just do it entirely on the Java side (serializing into a direct ByteBuffer) and without the intermediate conversion.

cataphract commented on libddwaf.cmake:
```diff
@@ -0,0 +1,33 @@
+set(LIBDDWAF_VERSION 1.16.0)
```
All right, but I'll need to update the toolchain to use c++20. Do you have any specific glibc version that you need to support? For instance, old versions of glibc don't support aligned_alloc, which libc++ uses (in its default configuration).

dmehala replied:
To my knowledge, NGINX relies on the version of glibc installed by the distribution or operating system. In theory, this shouldn't pose any issues. However, we distribute the module for certain variants such as Debian, Alpine, and AmazonLinux 3, the compatibility of the glibc version with these distributions could potentially become a bottleneck. I'll investigate 🕵️‍♂️

cataphract commented on src/datadog_context.cpp:
```diff
@@ -83,6 +93,13 @@ RequestTracing *DatadogContext::find_trace(ngx_http_request_t *request) {
   return nullptr;
 }
 
+RequestTracing &DatadogContext::single_trace() {
+  if (traces_.size() != 1) {
+    throw std::runtime_error{"Expected there to be exactly one trace"};
+  }
+  return traces_[0];
+}
```
That is not my understanding. We only care about main requests (the ones coming from outside), not subrequests. There can be multiple traces not if the worker is handling multiple requests (it generally is), but if there is a subrequest. The call to `single_trace` is made from here:

```
ngx_int_t on_access(ngx_http_request_t *request) noexcept try {
  if (request->main != request) {
    return NGX_DECLINED;
  }

  auto context = get_datadog_context(request);
  if (context == nullptr) {
    return NGX_DECLINED;
  }
  bool suspend = context->single_trace().on_main_request_start();
  if (suspend) {
    return NGX_AGAIN;
  }
  ...
```

so it checks if it's main request. Presumably, if we're handling the main request on `on_access`, there should be only one request, therefore only one trace (correct if I'm wrong).

dmehala replied:
Hmm you'r right. I made a bad assumption based on `DatadogContext` interface, my bad. However, `single_trace().on_main_request_start()` is still confusing. May I suggest `context->find_trace(request).on_main_request_start()`.

cataphract commented on src/datadog_directive.cpp:
```diff
@@ -1028,5 +1032,17 @@ char *hijack_auth_request(ngx_conf_t *cf, ngx_command_t *command,
 
+char *waf_thread_pool_name(ngx_conf_t *cf, ngx_command_t *command,
+                           void *conf) noexcept {
+  datadog_loc_conf_t *loc_conf = static_cast<datadog_loc_conf_t *>(conf);
```
We do not create thread pools explicitly. The user needs to create them. The reason for this is that there are important settings (the number of threads in the pool and the queue depth) that are better left to the user, because they interact with other user-controlled settings like the max number of workers. The configuration is like this:

```conf
thread_pool waf_thread_pool threads=2 max_queue=5;

http {
    datadog_waf_thread_pool_name waf_thread_pool;
}
```

you can create several pools and then use different ones from different locations, though I haven't tested that yet.

dmehala replied:
[NGINX Development guide](https://nginx.org/en/docs/dev/development_guide.html#threads) states:
> [...] calling ngx_thread_pool_add(cf, name), which either creates a new thread pool with the given name or returns a reference to the pool with that name if it already exists.

This approach can still inadvertently create a thread pool by directive per location. While the rationale behind is clear, there's a concern regarde its usage by customers who might not fully understand its implications, like, lots of "empty" thread pool (a large chunk of users copy/paste config). 

One way to address this concern is by creating a thread pool for WAF with an hardcoded name, such as `waf_thread_pool`. Savvy users , will still have the flexibility to configure it using `thread_pool waf_thread_pool threads=X max_queue=Y;`

That's just a suggestion.

cataphract replied:
I've changed it to `ngx_thread_pool_get`. This way the thread is not implicitly created and the user _must_ specify the thread pool to use (o/wise appsec will be disabled for that location).

It would arguably be more user friendly (though questionable for production purposes) to automatically create a thread pool, but we can leave such niceties for later.

cataphract commented on src/datadog_handler.cpp:
```diff
@@ -58,6 +58,27 @@ ngx_int_t on_enter_block(ngx_http_request_t *request) noexcept try {
+  if (context == nullptr) {
+    return NGX_DECLINED;
+  }
+  bool suspend = context->single_trace().on_main_request_start();
```
In the WAF, we're only interested in the main request, so this is actually better for us. If there were a `DataContext` for each request, we'd need to leave the WAF context empty for subrequests.

dmehala replied:
I spoke too fast... or not? I really need to take the take to understand that part of the code, so take the following with a grain of salt until then. 

My understanding right now is: there's actually one `DatadogContext` per request and attached to the request.

cataphract replied:
Yes, that's correct, and that's where I attached the security context now.

cataphract commented on src/datadog_handler.cpp:
```diff
@@ -71,5 +92,21 @@ ngx_int_t on_log_request(ngx_http_request_t *request) noexcept {
+  }
+
+  try {
+    return context->output_header_filter(*request);
```
what do you mean?

dmehala replied:
Sorry, just me ranting on the `DatadogContext` interface. The comment is mainly for me, so I can revisit the implementation.

cataphract commented on src/ngx_http_datadog_module.cpp:
```diff
@@ -308,6 +311,33 @@ static ngx_command_t datadog_commands[] = {
       0,
       NULL },
 
+    {
+      ngx_string("datadog_waf_thread_pool_name"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      waf_thread_pool_name,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(datadog_loc_conf_t, waf_thread_pool_name),
+      NULL
+    },
```
I've explain this in another comment. The thread pool is created by the user.

cataphract commented on src/security/blocking.cpp:
```diff
@@ -0,0 +1,401 @@
+namespace dns = datadog::nginx::security;
+
+struct block_resp {
+  enum class ct {
```
it does. I will rename it

cataphract commented on src/security/blocking.cpp:
```diff
@@ -0,0 +1,401 @@
+    return;
+  }
+  header->hash = 1;
+  header->key = ngx_stringv(name);
+  header->value = ngx_stringv(value);
```
I think it's fine. The key is a constant ("Location"sv), and the value comes from block_resp::location, which comes from block_spec::location, from resolve_block_spec, from create_block_request_action/create_redirect_request_action, from action_info::parameters (from its std::string). The action_info is stored in waf_handle::action_info_map_. The waf_handle has shared ownership in datadog::nginx::security::context, which is owned by RequestTracing, which is owned by DatadogContext::traces_... Ultimately, DatadogContext is only free on request cleanup.

dmehala replied:
Ok for the lifecyle but `lowercase_key` field is, from my understanding of https://www.nginx.com/resources/wiki/start/topics/examples/headers_management/ (at the end) is mandatory for proxy.

cataphract replied:
FWICT it's not relevant; these headers are to be sent to the client not to the web server behind the proxy in a reverse proxy configuration.

cataphract commented on src/datadog_context.cpp:
```diff
@@ -41,6 +42,15 @@ void DatadogContext::on_log_request(ngx_http_request_t *request) {
   trace->on_log_request();
 }
 
+ngx_int_t DatadogContext::output_header_filter(ngx_http_request_t &request) {
+  auto trace = find_trace(&request);
```
I generally prefer references to signal that the value can't be null, but I can change it if you prefer.

dmehala replied:
I agree 100% but consistency is key.

cataphract commented on src/request_tracing.h:
```diff
@@ -47,6 +53,7 @@ class RequestTracing {
   PropagationHeaderQuerier propagation_header_querier_;
   std::optional<dd::Span> request_span_;
   std::optional<dd::Span> span_;
+  security::context sec_ctx_;
```
Yeah, as I mentioned earlier, security::context should be on DatadogContext actually, because we're only interested in the main request.

cataphract replied:
Moved to DatadogContext

cataphract commented:
> I started to review 2-3 days ago under the assumption that the code was feature-complete. However, as I delved deeper, I found missing features, like HTTP/JSON template. It appears that new features are still being introduced, giving the impression that this should be treated as a draft PR.

Yes, I was doing some work and refactoring, but it's feature complete at this point (maybe I'll add the metrics, idk).

> * Thorough configuration testing: More extensive testing is necessary.

Yes, I'm not finished writing the tests.

> * `OWNERSHIP` file.

Will do.

> * Build
>   
>   * Introduce build argument to build either `Tracing` or `Tracing + ASM`.

Will do. See also my comment about updating the toolchain and the minimum glibc version.

>   * Generate a single `.so` for ease of use for the customer.
> * Enhanced observability: Add a span on `on_access` including details on the config file used, the rule executed, client ip and all WAF tags?

We have some standard metrics that track the time spent in the WAF, but I'm not sure I'll implement them in this first version. We also have reporting of client_ip and all the IP headers in some circumstances. That's also on the TODO list, but maybe not for now.

> 
> I refrained from reviewing anything under `securty/` because, ownership and potential extensive discussions on C++. Nevertheless, I recommend using comprehensible variable and function names and to stay consistent with the existing code.

Can you indicate the conventions you prefer? I can them create a .clang-tidy to enforce these.

> Lastly, it's worth noting that while you've successfully generalized Remote Config, its usage here introduces an avoidable dependency. We should consider delivering first WAF without Remote Config, then, adding this feature later.

I can create a version without remote config once I've addressed the comments here and added testing.

dmehala commented:
Cool, thanks.

I am on holiday and will be back April 10th, am I blocking you to move forward? If not, then I can revisit once I am back. My goal is to cut a new release and then we can start yoloing.

cataphract commented on src/security/blocking.cpp:
```diff
@@ -0,0 +1,401 @@
+    static bool is_more_specific(specificity a, specificity b) {
+      using underlying_t = std::underlying_type_t<specificity>;
+      return static_cast<underlying_t>(a) > static_cast<underlying_t>(b);
+    }
```
I've renamed it `first_is_more_specific` to remove the ambiguity.

cataphract commented:
> I am on holiday and will be back April 10th, am I blocking you to move forward? If not, then I can revisit once I am back. My goal is to cut a new release and then we can start yoloing.

The 10th is fine. This week we're working on random stuff (aka "Inovation Week") so I'm pausing work on this.

Anilm3 commented on src/security/util.h:
```diff
@@ -0,0 +1,172 @@
+
+struct NgxStrEqual {
+  bool operator()(const ngx_str_t &lhs, const ngx_str_t &rhs) const {
+    if (lhs.len != rhs.len) {
+      return false;
+    }
+    return std::memcmp(lhs.data, rhs.data, lhs.len) == 0;
```
```suggestion
    return hs.len == rhs.len && std::memcmp(lhs.data, rhs.data, lhs.len) == 0;
```

Anilm3 commented on src/security/util.h:
```diff
@@ -0,0 +1,172 @@
+    return *this;
+  }
+
+  T &operator*() const { return elts_[index_]; }
```
Should check and throw if `elts_` is `nullptr`

cataphract replied:
Generally iterators behave like pointers and don't check invalid dereferences.

Anilm3 commented on src/security/ddwaf_obj.cpp:
```diff
@@ -0,0 +1,95 @@
+namespace datadog::nginx::security {
+namespace impl {
+template <typename D>
+void json_to_obj_impl(DdwafMemres &memres, ddwaf_obj &object, const D &doc) {
```
Actually this one probably should take an optional max depth just in case we decide to use it to parse json bodies. The ruleset generation should not be limited in depth.

cataphract replied:
This is used only to parse configuration. But I'm adding a limit anyway.

Anilm3 commented on nginx-module.cmake:
```diff
@@ -17,31 +17,46 @@ list(REMOVE_DUPLICATES NGINX_CONF_ARGS)
-  PRIVATE
-    ${NGINX_MODULE_FILE}
-)
+set(NGINX_VERSION "" CACHE STRING "The nginx version")
```
This could have a default value as well, in which case a warning would be sufficient. Even better if it were possible to download the latest stable version although unclear how that could be determined.

cataphract replied:
Perhaps, but that is not the current behavior. You need to see NGINX_VERSION.

dmehala commented on nginx-module.cmake:
```diff
@@ -17,31 +17,46 @@ list(REMOVE_DUPLICATES NGINX_CONF_ARGS)
-    ${NGINX_SRC_DIR}/src/event/modules
-    ${NGINX_SRC_DIR}/objs
-)
+FetchContent_Declare(
```
`NGINX_SRC_DIR` is the way. Allow me to copy/paste a previous comment:
> Now, CMake expects the NGINX source directory as input, specified with `-DNGINX_SRC_DIR`, and configures accordingly. Instead of downloading the source code each time, which isn't ideal given that I work with both the stable and mainline NGINX branch, there are a few advantages to this approach:
>
>1. It's quicker to simply rename the folder or adjust the project configuration to use a different directory rather than waiting for CMake to download the source tarball (unless CMake has some caching mechanism).
>2. By avoiding unnecessary downloads every-time consider it should re-configure the project, I'm contributing to saving the planet (unless CMake has some caching mechanism).
>3. I'm not aware of a .tar.gz file following the HEAD of NGINX.

Anilm3 replied:
@dmehala, we spoke offline and agreed to have both solutions coexist, so your workflow should be unaffected.

dmehala replied:
Why should we have two solution for one problem? As a user that don't know anything about the module, that will be confusing.

Anilm3 replied:
Having two solutions for this specific problem is not a concern if it enhances the developer experience and lets each developer adjust their workflow to what best suits them.

> As a user that don't know anything about the module, that will be confusing. 

This is only a problem as long as we don't document both methods adequately and ensure at least one of them is required by default.

> It's added complexity for no benefit.

I don't think it's much complexity at all and the benefit is that users and developers won't need to specifically have the nginx source locally and can easily choose the version they prefer with a cmake variable. Bear in mind this is the way most build systems work today in most languages so it's a mischaracterization to say there is no benefit.

In any case, as mentioned, having both **mutually exclusive** options will allow you to continue using your preferred workflow, but to answer some of your points:

> By avoiding unnecessary downloads every-time consider it should re-configure the project, I'm contributing to saving the planet (unless CMake has some caching mechanism).

The download of the source only needs to happen once, further downloads or updates can be inhibited using `FETCHCONTENT_UPDATES_DISCONNECTED`, more info [here](https://cmake.org/cmake/help/latest/module/FetchContent.html#variable:FETCHCONTENT_UPDATES_DISCONNECTED).

> I'm not aware of a .tar.gz file following the HEAD of NGINX.

For HEAD one can still use `NGINX_SRC_DIR`.

Anilm3 commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,2 @@
+* @DataDog/apm-proxy
+src/security/ @DataDog/asm-libddwaf
```
We now have at least write access, which should be sufficient for now.

dmehala commented on src/recommended.json:
expected?

dmehala commented on src/request_tracing.cpp:
can revert the whole file

cataphract commented on src/datadog_conf.h:
```diff
@@ -116,6 +116,39 @@ struct datadog_main_conf_t {
+
+  // DD_APPSEC_WAF_TIMEOUT (default: 0.1 s), in microseconds
+  // The default value is not set to 100 to detect when the value is unset
+  ngx_msec_t appsec_waf_timeout_ms{NGX_CONF_UNSET_MSEC};
```
I've clarified the comment.

dmehala replied:
- [ ] Update comment

cataphract commented on src/security/ddwaf_obj.h:
```diff
@@ -0,0 +1,381 @@
+
+  template <typename T>
+  ddwaf_obj &make_number(T value) {
+    static_assert(std::is_arithmetic_v<T>, "T must be an arithmetic type");
```
Right, I'm targeting c++17. Although I think this is better, because you get the failed assertion rather than an error about no method found.

cataphract commented on src/security/ddwaf_obj.h:
```diff
@@ -0,0 +1,381 @@
+  }
+
+  template <typename T = ddwaf_obj>
+  T &at_unchecked(nb_entries_t index) const {
```
Yeah, I think I prefer this to operator overloading because operator[] as used for maps (which ddwaf objects can be) has questionable semantics.

cataphract commented on src/security/ddwaf_obj.h:
```diff
@@ -0,0 +1,381 @@
+    return at_unchecked<T>(index);
+  }
+
+  std::size_t size() const { return nbEntries; }
```
I'll return nb_entries_t instead.

cataphract commented on src/security/ddwaf_obj.h:
```diff
@@ -0,0 +1,381 @@
+  }
+
+  template <typename T = ddwaf_obj>
+  T get(std::string_view key) const {
```
I think I prefer to use `get` to fetch a key, and `at` to fetch the map entry on a numeric index (which mirrors `at` on arrays).

cataphract commented on src/security/ddwaf_obj.h:
```diff
@@ -0,0 +1,381 @@
+  }
+
+  template <typename T = ddwaf_obj>
+  T &get_entry_unchecked(std::size_t i) {
```
I'm changing it to `at_unchecked`, as mentioned above.

cataphract commented on src/security/util.h:
```diff
@@ -0,0 +1,172 @@
+  }
+
+  for (std::size_t i = 0; i < header.key.len; ++i) {
+    if (std::tolower(header.key.data[i]) != key[i]) {
```
I prefer to avoid the runtime overhead versus the the potential of a bug. I'll add a check on debug builds.

cataphract commented:
@dmehala I see you have a branch for cpp-20... I'm doing work that overlaps a bit, with musl (glibc-compatible)/libc++. That way we don't need separate builds for alpine

Anilm3 commented on src/security/blocking.cpp:
```diff
@@ -0,0 +1,409 @@
+    int status;
+    enum ContentType ct;
+
+    status = spec.status;
```
nit: this seems pointless.

Anilm3 commented on src/security/blocking.cpp:
```diff
@@ -0,0 +1,409 @@
+
+  ngx_http_discard_request_body(&req);
+
+  // TODO: clear all current headers?
```
This is important if feasible...

cataphract replied:
Because we don't block on response, the problem becomes moot.

Anilm3 commented on src/security/blocking.cpp:
```diff
@@ -0,0 +1,409 @@
+  if (!resp.location.empty()) {
+    push_header(req, "Location"sv, resp.location);
+  }
+  if (templ) {
```
I wouldn't expect there to be a template if we're redirecting (maybe an empty one?), perhaps the logic could be split to better reflect the actions taken when redirecting vs when overriding the body.

cataphract replied:
Technically, there's nothing with sending a response entity on a redirect. In fact, it's very common. While we don't currently send anything, I don't think this method needs to reflect that.

Anilm3 commented on src/security/client_ip.cpp:
```diff
@@ -0,0 +1,601 @@
+      out.af = AF_INET;
+      return {out};
+    }
+  }
```
Is it intentional to ignore the hint if `inet_pton` fails? The logic then differs when the hint is incorrect for IPv6

cataphract replied:
if it fails, it might still be an ipv6-mapped ipv4 address, which it why it continues. However, from the usage isn't very accurate, because it calls with AF_INET6 if it within brackets. For that what matters is that it's formally ipv6. I'll adjust.

cataphract commented on src/glibc_compat.c:
```diff
@@ -0,0 +1,117 @@
+#if defined(__linux__) && !defined(__GLIBC__)
+
+#ifdef __x86_64__
+float ceilf(float x) {
```
This is for both glibc and musl, when compiled with musl.

cataphract commented on src/ngx_http_datadog_module.cpp:
```diff
@@ -461,7 +568,18 @@ static ngx_int_t datadog_init_worker(ngx_cycle_t *cycle) noexcept try {
   }
 
-  auto maybe_tracer = TracingLibrary::make_tracer(*main_conf);
+  std::shared_ptr<dd::Logger> logger = std::make_shared<NgxLogger>();
```
ah it's there because the version supporting remote config does use it

cataphract commented on src/ngx_http_datadog_module.cpp:
```diff
@@ -394,24 +482,36 @@ static ngx_int_t datadog_master_process_post_config(
-    if (const char *value = std::getenv(name.c_str())) {
-      main_conf->environment_variables.push_back(
-          environment_variable_t{.name = name, .value = value});
+  auto push_to_main_conf = [main_conf](const std::string &env_var_name) {
+    if (const char *value = std::getenv(env_var_name.c_str())) {
+      main_conf->environment_variables.push_back(environment_variable_t{
+          .name = std::move(env_var_name), .value = value});
```
The copy of the std::string is usually not needed (only if the env variable is set), which is why it's passed by reference. But yes, the `std::move` is redundant for when the copy is needed (it will call the copy constructor in both cases).

Anilm3 replied:
On the other hand you're creating a copy of the original `string_view` into a temporary, passing it as a const reference and then copying it again when the environment variable exists. A concern could be around the nul-termination of those views, as they need to be passed to `std::getenv`, however in both cases (tracing and appsec), they are originally literals. Either way, I suppose this is a safer approach if you believe the source of those names could eventually lack nul-termination, otherwise it might make more sense to just pass a `string_view`.

cataphract replied:
Ah I should've checked the caller. You're right, it makes more sense to take an `std::string`.

Anilm3 commented on src/security/collection.cpp:
```diff
@@ -0,0 +1,396 @@
+      } else {
+        // subsequent occurrence of this key
+        auto &arr_val = *ie->second;
+        if (!it.is_delete()) {
```
What does this mean? When would `is_delete` be true?

cataphract replied:
modules can delete existing headers. For this, they add a header with hash = 0 (and add headers with hash 1; hence "hash" doesn't represent a hash in response headers like it does in request headers, even though they use the same data structure)

Anilm3 replied:
Ah that makes sense, thanks for explaining.

Anilm3 commented on src/security/collection.cpp:
```diff
@@ -0,0 +1,396 @@
+    set_map_entry_str(slot, kMethod, request.method_name);
+  }
+
+  // adapt to the same iteratror format as query_string_iter
```
```suggestion
  // adapt to the same iterator format as query_string_iter
```

Anilm3 commented on src/security/collection.cpp:
```diff
@@ -0,0 +1,396 @@
+    HeaderKeyValueIter &operator++() {
+      while (true) {
+        ++it_;
+        if (!(it_ != end_)) {
```
```suggestion
        if (it_ == end_) {
```
The operator exists, better hide the double negative.

Anilm3 commented on src/security/collection.cpp:
```diff
@@ -0,0 +1,396 @@
+
+    void reset() { it_ = decltype(it_){list_}; }
+
+    bool ended() const noexcept { return !(it_ != end_); }
```
```suggestion
    bool ended() const noexcept { return it_ == end_; }
```

Anilm3 commented on src/security/collection.cpp:
```diff
@@ -0,0 +1,396 @@
+        return false;
+      } else {  // response headers
+        const auto &h = *it_;
+        return h.hash == 0;
```
Still puzzled of what this means... can't the iterator skip this entry during `operator++`?

cataphract replied:
It can't be skipped because it signals important information (that the header previously defined should be deleted). It _could_ be skipped only if all the references to the header up to the _last_ deletion would be deleted as well. But that would require a first pass and saving this information somewhere.

Anilm3 commented on src/security/context.cpp:
```diff
@@ -0,0 +1,672 @@
+  std::string_view const json{buffer.GetString(), buffer.GetLength()};
+
+  ngx_str_t json_ns{dnsec::ngx_stringv(json)};
+  ngx_log_error(NGX_LOG_NOTICE, req.connection->log, 0, "appsec event: %V",
```
Any particular reason to choose notice instead of debug? I'm not sure what the typical setup will be but this is a pretty verbose log statement.

cataphract replied:
Notice is between info and warn. So maybe info is better instead? Debug is not a great option, because it's compiled to nothing by default. You need to configure nginx with --debug and then build the extension against that, which we don't do (and I doubt the result would be binary compatible, though I didn't check).

cataphract replied:
changed to info

Anilm3 commented on src/security/ddwaf_memres.h:
```diff
@@ -0,0 +1,99 @@
+  ~DdwafMemres() = default;
+
+  template <typename T = ddwaf_object>
+  T *allocate_objects(std::size_t num_objects) {
```
Why is this method templated? Is it to alias to one of your object variants? Would be nice to have a comment somewhere, rationale isn't immediately obvious.

cataphract replied:
Yes, that's the idea, as reflected in the static assertions (T must be a subclass of ddwaf_object, with same size and alignment and must satisfy the standard layout requirements), although I see calling code only uses ddwaf_object and `dnsec::ddwaf_obj`, so I could also return the latter.

Anilm3 commented on src/security/library.cpp:
```diff
@@ -0,0 +1,585 @@
+static constexpr ddwaf_config kBaseWafConfig{
+    .limits =
+        {
+            .max_container_size = 150,
+            .max_container_depth = 10,
```
Defaults are typically 256 and 20 respectively, why did you choose these?

cataphract replied:
changed to those

Anilm3 commented on src/security/library.cpp:
```diff
@@ -0,0 +1,585 @@
+
+namespace {
+
+static constexpr auto kDefaultKeyRegex =
```
There is a default defined here and one in `FinalizedConfigSettings`, you could get rid of one probably.

cataphract replied:
Indeed. I also see that I considered emptiness and non-presence of the environment variable the same, which shouldn't be done.

Anilm3 commented on src/security/library.cpp:
```diff
@@ -0,0 +1,585 @@
+      return DDWAF_LOG_DEBUG;
+    case NGX_LOG_INFO:
+    case NGX_LOG_NOTICE:
+      return DDWAF_LOG_INFO;
```
Some libraries are setting WAF INFO logs to DEBUG as they can be noisy and provide little value to the customer.

cataphract replied:
see my comment before about debug logs

dmehala reviewed (commented):
There's a few questions there and there, overall you did a good job. 

You greatly improved our CI, which is something I wanted to do since I joined but never really got the time/invested enough time to do it. Thank you.

There's still a pending question regarding which version of dd-trace-cpp should be used. We agreed the first iteration with ASM should not use Remote Config, thus, dd-trace-cpp should point to the latest version `v0.2.0`.

Additionally, the main README should mention ASM is supported, how to build with ASM and also `configuration.md` should have a dedicated section introducing new ASM directives.

dmehala commented on src/ngx_http_datadog_module.cpp:
```diff
@@ -461,7 +568,18 @@ static ngx_int_t datadog_init_worker(ngx_cycle_t *cycle) noexcept try {
+  } catch (const std::exception &e) {
+    ngx_log_error(NGX_LOG_ERR, cycle->log, 0,
+                  "Initialising security library failed: %s", e.what());
+    return NGX_ERROR;
```
I would argue it should not prevent tracing from working.

Anilm3 replied:
We have two options:
- Since this is going to be a beta initially, we could fail for now since the original module containing only tracing is still a separate one.
- We could override the relevant `enabled` configuration to `off` and allow tracing to continue.

Either way, in my view this isn't a blocker for now, but we should definitely reevaluate before GA.

dmehala commented on src/datadog_context.cpp:
```diff
@@ -32,12 +41,47 @@ void DatadogContext::on_change_block(ngx_http_request_t *request,
     throw std::runtime_error{
         "on_log_request failed: could not find request trace"};
   }
+#ifdef WITH_WAF
+  if (sec_ctx_ && request == request->main) {
+    sec_ctx_->on_main_log_request(*request, trace->active_span());
+  }
+#endif
```
I would argue this should be executed after tracing. Our goal is to provide "precise" timing to our customer. What do you think?

cataphract replied:
Logging happens after the request is fully served, so I guess it would make more sense to move it after tracing, yes. This is time spent on the main thread that will work towards increasing the latency and total processing time of other requests, but that would happen the same way regardless of the location of the call to `on_main_log_request`.

labbati commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,2 @@
+* @DataDog/apm-proxy
```
I don't think so. For `*` I think we should have `c++` contributors from both language platform and SDK.

cataphract commented on src/security/collection.cpp:
```diff
@@ -0,0 +1,396 @@
+
+    // we now know the number of keys; allocate map entries
+    dnsec::ddwaf_obj *entries =
+        memres_.allocate_objects<dnsec::ddwaf_obj>(keys_bag.size());
```
I thought duplicates were tolerated, not fully supported. I mean, how do you even indicate paths of offending data? a > b` is now ambiguous if `a` contains several entries keyed `b`.

But anyway, a first pass is always necessary to avoid reallocations further down the road.

cataphract commented on bin/release.py.new.unmerged:
So. This file is my version of `release.py` before you changed `released.py` to replace quoting style and other stuff that caused conflicts. Rather than trying to resolve the conflicts, I'd prefer that you review this file to see if you agree with the naming given to the artifacts. Anil already said that WAF shouldn't be used, but rather "appsec". The names also change in two other ways: they're not image specific, the only thing that matters is version and architecture, so the name of the image is stripped from the name. Also I'm uploading the debug symbols as well,.

dmehala replied:
I resolved the conflict in e361779b3e80ba7da402bb36af250dd21766680f

Before GA I think it's worth revisiting if we should produce two versions of the module (TRACING and TRACING + APPSEC) or one module with all products but with good default and the possibility to disable/enable per products. What do you think @Anilm3 @cataphract ?

cataphract commented on .circleci/config.yml:
```diff
@@ -133,1013 +141,451 @@ jobs:
     - checkout
-    - run: find bin/ test/ example/ -type f -executable | xargs shellcheck --exclude SC1071,SC1091,SC2317
-
+    - run: find bin/ test/ example/ -type f -executable | xargs shellcheck --exclude
+        SC1071,SC1091,SC2317
```
yaml is a funny format; this is actually just a folded representation of a string without newlines.

cataphract commented on .circleci/config.yml:
```diff
@@ -1,129 +1,137 @@
+    - checkout
+    - run: git submodule sync && git submodule update --init --recursive
+    - run:
+        command: 'make build-musl'
```
It builds against musl in a linux VM. I think "build" should be reserved for using the toolchain on the host machine. It would be strange if "make build" would generate a library for Linux when run on Mac OS.

dmehala replied:
Wow I was not focused when I wrote that. I meant, now, musl is the default libc?

cataphract commented on .circleci/config.yml:
I have. I then generated this [draft release](https://github.com/DataDog/nginx-datadog/releases/tag/untagged-aa7b821fde252827b9e8) with the `release.py.new.unmerged` script

dmehala replied:
Amazing, I was afraid you'll encounter an issue on older versions. Thank you for simplifying that mess. 🙏

cataphract commented on CMakeLists.txt:
```diff
@@ -1,8 +1,10 @@
-# The C code doesn't care about the C++ version.
-# dd-trace-cpp works with C++17 or newer.
-set(CMAKE_CXX_STANDARD 17)
+cmake_minimum_required(VERSION 3.24)
```
It's because of `ZLIB_USE_STATIC_LIBS`. It was added in 3.24.

cataphract commented on bin/generate_jobs_yaml.rb:
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env ruby
+
```
It has a bang line, so it's run the usual fashion (provided `ruby` exists in the PATH). As for the description, there was already one in `bin/README.md`.

cataphract commented on dd-trace-cpp:
I'll revert the reference to v0.2.0. I must have accidentally committed the update.

cataphract replied:
Actually the main branch because I need need methods added by Anil.

Anilm3 reviewed (approved):
No further concerns from me, aside from:
-  The unmerged script that needs to be reconciled, I think we can move on once @dmehala is satisfied. 
- There should be a small section in the README.md explaining the configuration changes required to get ASM working, with an example.

Overall, fantastic work as usual, my tests so far, while not exhaustive, have worked as expected, although the subsequent system tests PR will be exhaustive enough.

dmehala reviewed (approved):
Amazing work! :shipit:

cataphract commented:
@dmehala 

> * X module per product or one big module with all products?

I think we can leave two modules for now. The thing I'm more afraid of is that someone may have a build without `--threads` and that the module fails to load due failing symbol resolution.

> * Update public documentation (new artifact scheme).

I'm going to open a PR there.

> * Add a simple example in `example` folder.

I've updated the example.

> * Small section in the `README.md` explaining the configuration changes required to get ASM working.

Done as well.

I also found some memory corruption doing benchmarking that I've addressed in https://github.com/DataDog/nginx-datadog/pull/65/commits/8a22ddf80696f0e5475d402a7facc7f83750e628 . I'm not 100% sure it's the best approach (as opposed to changing `connection->read/write->handler`) but we'll see.
